core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -4548,22 +4562,59 @@ private void analyzeRowFilter(String currentIdentity, Table table, QualifiedObje
                 throw new TrinoException(INVALID_ROW_FILTER, extractLocation(table), format("Row filter for '%s' is recursive", name), null);
             }
 
-            Expression expression;
-            try {
-                expression = sqlParser.createExpression(filter.getExpression(), createParsingOptions(session));
-            }
-            catch (ParsingException e) {
-                throw new TrinoException(INVALID_ROW_FILTER, extractLocation(table), format("Invalid row filter for '%s': %s", name, e.getErrorMessage()), e);
-            }
+            Expression expression = parseRowConstraintExpression(table, name, filter, INVALID_ROW_FILTER, "Invalid row filter");
 
             analysis.registerTableForRowFiltering(name, currentIdentity);
 
             verifyNoAggregateWindowOrGroupingFunctions(session, metadata, expression, format("Row filter for '%s'", name));
 
             ExpressionAnalysis expressionAnalysis;
             try {
-                expressionAnalysis = ExpressionAnalyzer.analyzeExpression(
-                        createViewSession(filter.getCatalog(), filter.getSchema(), Identity.ofUser(filter.getIdentity()), session.getPath()), // TODO: path should be included in row filter
+                expressionAnalysis = getExpressionAnalysis(table, name, scope, filter, expression, "Invalid row filter");
+            }
+            finally {
+                analysis.unregisterTableForRowFiltering(name, currentIdentity);
+            }
+
+            analysis.recordSubqueries(expression, expressionAnalysis);
+
+            addCoercionForRowConstraint(table, name, expression, expressionAnalysis, "row filter");
+
+            analysis.addRowFilter(table, expression);
+        }
+
+        private void analyzeCheckConstraint(Table table, QualifiedObjectName name, Scope scope, ViewExpression filter)
```

Comment:
We need to ensure the expression is "retrospectively deterministic" as describe by the SQL spec. By that, it means that:

> Certain `<boolean value expression>`s are identified as “retrospectively deterministic”. A retrospectively deter- ministic `<boolean value expression>` has the property that if it is True at one point time, then it is True for all later points in time if re-evaluated for the identical SQL-data by an arbitrary user with the identical set of privileges. 

It's hard to determine if an expression is retrospectively deterministic, but we should be conservative and ensure it, at least, always deterministic.

Reply:
Added a check to verify that an expression doesn't contain non-deterministic functions. Let me know if you have additional suggestions.